### PR TITLE
Mac support

### PR DESCRIPTION
**Pull request by jvns**
Preliminary Mac support! In my testing so far, this works with Ruby (as long as it's installed by rbenv) on Mac.

This does **not** work with system Ruby on Mac. There are a couple issues with adding system Ruby support:

1. SIP -- getting a port to talk to system Ruby is impossible if you have SIP enabled on your Mac (tested this today). I believe this isn't fixable, the whole point of SIP is to prevent other programs from instrumenting your program in exactly this way.
2. my implementation of `get_process_maps` right now doesn't support dynamically loaded libraries, and all the symbols in system Ruby (at least on the High Sierra mac I'm working with) are in a dynamically loaded library which I haven't been about to find the address or path of yet. more notes about the difficulties there in [this blog post I wrote today](https://jvns.ca/blog/2018/01/26/mac-memory-maps/)

issue 2 there is definitely possible to work around but I think it might take me another 2 days so I think it's probably better to merge this as is. (and the SIP issue makes me wonder if it's even worth trying to support system Ruby, my guess is most people have SIP enabled)

Still need to add CI for Mac but wanted to get this out there.

### CODE REVIEW

**Comment by jvns** [comment]
I think this is working now! I added a little bit of CI for mac (no integration tests still, but I fought with travis + rvm + rbenv those for 4 hours and didn't get it to work so giving up on this for now).

I tested this with both rvm and rbenv on my Mac dev machine and it worked. It still doesn't work with system Ruby and I'm not planning system Ruby support for the time being.

**Comment by parkr** [comment]
@jvns Well done!!! ✨ 🌟 To test, pull, recompile, and run `sudo ~/path/to/rbspy <command>`? I didn't have the requirement to run sudo before on Mac. Is that something that can be caught?

**Comment by jvns** [comment]
@parkr I started requiring sudo just because I didn't think it was possible to make it work without sudo (it didn't work on my dev machine). did it work for you without sudo for you in the past If so I'll remove the requirement and adjust the error messages.

**Comment by parkr** [comment]
It's not hanging anymore, but I'm still not having success! Does any of this help you?

```text
~/code/rbspy#mac-support$ rm -rf target && cargo build
~/code/rbspy#mac-support$ sudo --preserve-env ./target/debug/rbspy record ruby ./ci/ruby-programs/infinite.rb Dropping permissions: running Ruby command as user parkr
Error. Causes:
- Couldn't determine Ruby version
- Couldn't connect to PID
- File exists (os error 17)

~/code/rbspy#mac-support$ rbenv version
2.4.3 (set by /Users/parkr/code/rbspy/.ruby-version)
~/code/rbspy#mac-support$ sudo --preserve-env env RUST_LOG=debug RUST_BACKTRACE=1 ./target/debug/rbspy record ruby ./ci/ruby-programs/infinite.rb
Dropping permissions: running Ruby command as user parkr
Error. Causes:
- Couldn't determine Ruby version
- Couldn't connect to PID
- File exists (os error 17)
stack backtrace:
   0:        0x10fd4ab2c - backtrace::backtrace::trace<closure>
                        at /Users/parkr/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.5/src/backtrace/mod.rs:42
   1:        0x10fd4210f - backtrace::capture::Backtrace::new_unresolved::h22a3194b78596c3b
                        at /Users/parkr/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.5/src/capture.rs:88
   2:        0x10fd3ffc1 - failure::backtrace::internal::{{impl}}::new
                        at /Users/parkr/.cargo/registry/src/github.com-1ecc6299db9ec823/failure-0.1.1/src/backtrace/internal.rs:44
   3:        0x10fd404fe - failure::backtrace::Backtrace::new::h3969a68773a2cd75
                        at /Users/parkr/.cargo/registry/src/github.com-1ecc6299db9ec823/failure-0.1.1/src/backtrace/mod.rs:98
   4:        0x10f802540 - failure::error::{{impl}}::from<std::io::error::Error>
                        at /Users/parkr/.cargo/registry/src/github.com-1ecc6299db9ec823/failure-0.1.1/src/error.rs:34
   5:        0x10f7ff4e9 - core::convert::{{impl}}::into<std::io::error::Error,failure::error::Error>
                        at /Users/travis/build/rust-lang/rust/src/libcore/convert.rs:415
   6:        0x10f7de128 - failure::context::{{impl}}::with_err<&str,std::io::error::Error>
                        at /Users/parkr/.cargo/registry/src/github.com-1ecc6299db9ec823/failure-0.1.1/src/context.rs:78
   7:        0x10f803dec - failure::Fail::context<std::io::error::Error,&str>
                        at /Users/parkr/.cargo/registry/src/github.com-1ecc6299db9ec823/failure-0.1.1/src/lib.rs:127
   8:        0x10f7f02bc - failure::result_ext::{{impl}}::context::{{closure}}<u32,std::io::error::Error,&str>
                        at /Users/parkr/.cargo/registry/src/github.com-1ecc6299db9ec823/failure-0.1.1/src/result_ext.rs:32
   9:        0x10f859d0d - core::result::{{impl}}::map_err<u32,std::io::error::Error,failure::context::Context<&str>,closure>
                        at /Users/travis/build/rust-lang/rust/src/libcore/result.rs:500
  10:        0x10f850b10 - failure::result_ext::{{impl}}::context<u32,std::io::error::Error,&str>
                        at /Users/parkr/.cargo/registry/src/github.com-1ecc6299db9ec823/failure-0.1.1/src/result_ext.rs:32
  11:        0x10f7edd55 - rbspy::initialize::get_ruby_version_retry
                        at src/initialize.rs:106
  12:        0x10f7ecf02 - rbspy::initialize::initialize
                        at src/initialize.rs:29
  13:        0x10f8399e5 - rbspy::record
                        at src/main.rs:208
  14:        0x10f838698 - rbspy::do_main
                        at src/main.rs:137
  15:        0x10f838ea3 - rbspy::main
                        at src/main.rs:168
  16:        0x10fd76e1e - __rust_maybe_catch_panic
                        at src/libpanic_unwind/lib.rs:101
  17:        0x10fd6c278 - std::rt::lang_start::h0ff358a55ab2ec4c
                        at src/libstd/rt.rs:58
  18:        0x10f83d644 - main
```

**Comment by jvns** [comment]
Wow, I have no idea what "os error 17" means --  that's a new one. This is all I've been able to find out about it so far:

```
$ macerror 17
Mac OS error 17 (dsNoPackErr): package 0 not present
```

File: src/address_finder.rs [hunk 1]
```diff
@@ -14,54 +14,128 @@ pub enum AddressFinderError {
     #[fail(display = "No process with PID: {}", _0)] NoSuchProcess(pid_t),
     #[fail(display = "Permission denied when reading from process {}. Try again with sudo?", _0)]
     PermissionDenied(pid_t),
+    #[fail(display = "Couldn't get port for PID {}. Possibilities: that process doesn't exist or you have SIP enabled and you're trying to profile system Ruby (try rbenv instead).", _0)]
+    MacPermissionDenied(pid_t),
```
Comment (kamalmarhubi) [inline src/address_finder.rs]: not 100% on this error variant's existence. at least `#[cfg(target_os = "macos")]` it.

File: src/mac_maps.rs [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        } else {
+            continue;
+        };
+        vec.push(sym);
+    }
+    vec
+}
+
+pub fn get_symbols(filename: &str) -> Result<Vec<Symbol>, Error> {
+    let output = std::process::Command::new("nm").arg(filename).output()?;
```
Comment (kamalmarhubi) [inline src/mac_maps.rs]: will people without xcode tools have `nm`?

File: src/mac_maps.rs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    pub fn is_write(&self) -> bool {
+        self.info.protection & mach::vm_prot::VM_PROT_WRITE != 0
+    }
+    pub fn is_exec(&self) -> bool {
+        self.info.protection & mach::vm_prot::VM_PROT_EXECUTE != 0
+    }
+}
+
+pub fn get_process_maps(pid: pid_t, task: mach_port_name_t) -> Vec<MacMapRange> {
+    let init_region = mach_vm_region(pid, task, 1).unwrap();
```
Comment (kamalmarhubi) [inline src/mac_maps.rs]: why 1 and not 0?

File: src/mac_maps.rs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    }
+    pub fn is_write(&self) -> bool {
+        self.info.protection & mach::vm_prot::VM_PROT_WRITE != 0
+    }
+    pub fn is_exec(&self) -> bool {
+        self.info.protection & mach::vm_prot::VM_PROT_EXECUTE != 0
+    }
+}
+
+pub fn get_process_maps(pid: pid_t, task: mach_port_name_t) -> Vec<MacMapRange> {
```
Comment (kamalmarhubi) [inline src/mac_maps.rs]: some comment here explaining how the `mach_vm_region` api works would be good :-)

File: src/mac_maps.rs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        start: address,
+        count: count,
+        filename: filename,
+    })
+}
+
+pub fn task_for_pid(pid: pid_t) -> io::Result<mach_port_name_t> {
+    let mut task: mach_port_name_t = MACH_PORT_NULL;
+    // sleep for 5ms to make sure we don't get into a race between `task_for_pid` and execing a new
+    // process. Races here can freeze the OS because of a Mac kernel bug.
```
Comment (kamalmarhubi) [inline src/mac_maps.rs]: link to your blag

File: src/main.rs [hunk 6]
```diff
@@ -119,7 +145,26 @@ fn do_main() -> Result<(), Error> {
...
+        println!("rbspy only works as root on Mac. Try rerunning with `sudo --preserve-env !!`.");
+        println!(
+            "If you run `sudo rbspy record ruby your-program.rb`, rbspy will drop privileges when running `ruby your-program.rb`. If you want the Ruby program to run as root, use `rbspy --no-drop-root`."
+        );
+        return false;
+    }
+}
+
 fn main() {
+    if cfg!(target_os = "macos") {
```
Comment (kamalmarhubi) [inline src/main.rs]: alternatively, have a `#[cfg(target_os = "macos")]` and `#[cfg(not(target_os = "macos"))]` version of `sudo_if_not_root`, where the second just returns true. maybe change the name in that case though.

File: src/mac_maps.rs (hunk 2 above)
Comment (jvns) [reply src/mac_maps.rs]: Not sure! I couldn't get the Mach-O symbol-parsing code from goblin to work (I think there's a bug in it), which is why I used `nm`. Need to report a bug in goblin.

**Comment by jvns** [comment]
I'm going to merge this as-is to get preliminary Mac support into master, and then make a separate issue to track @parkr's bug and switch away from nm since I do want to figure that out.

**Pull request merged**